packages/passport/sdk/src/starkEx/workflows/workflows.test.ts

```diff
@@ -0,0 +1,105 @@
+import { ImxApiClients, imx } from '@imtbl/generated-clients';
+import {
+  Environment,
+  IMXClient,
+  ImmutableConfiguration,
+  NftTransferDetails,
+  UnsignedExchangeTransferRequest,
+  UnsignedOrderRequest,
+  UnsignedTransferRequest,
+} from '@imtbl/x-client';
+import AuthManager from 'authManager';
+import { PassportError, PassportErrorType } from 'errors/passportError';
+import GuardianClient from 'guardian';
+import MagicAdapter from 'magicAdapter';
+import { PassportImxProvider } from 'starkEx/passportImxProvider';
+import { PassportEventMap, PassportEvents } from 'types';
+import TypedEventEmitter from 'utils/typedEventEmitter';
+
+describe('passportImxProvider auth tests', () => {
+  const mockAuthManager = {
+    login: jest.fn(),
+    getUser: jest.fn(),
+    forceUserRefresh: jest.fn(),
+  };
+
+  const immutableXClient = new IMXClient({
+    baseConfig: new ImmutableConfiguration({
+      environment: Environment.SANDBOX,
+    }),
+  });
+
+  const passportEventEmitter = new TypedEventEmitter<PassportEventMap>();
+
+  const magicAdapterMock = {
+    login: jest.fn(),
+  };
+
+  const imxApiClients = new ImxApiClients({} as any);
+
+  const mockGuardianClient = {
+    withDefaultConfirmationScreenTask: (task: () => any) => task,
+    withConfirmationScreenTask: () => (task: () => any) => task,
+  };
+
+  const passportImxProvider = new PassportImxProvider({
+    authManager: mockAuthManager as unknown as AuthManager,
+    magicAdapter: magicAdapterMock as unknown as MagicAdapter,
+    guardianClient: mockGuardianClient as unknown as GuardianClient,
+    immutableXClient,
+    passportEventEmitter,
+    imxApiClients,
+  });
+
+  describe.each([
+    ['transfer' as const, {} as UnsignedTransferRequest],
+    ['createOrder' as const, {} as UnsignedOrderRequest],
+    ['cancelOrder' as const, {} as imx.GetSignableCancelOrderRequest],
+    ['createTrade' as const, {} as imx.GetSignableTradeRequest],
+    ['batchNftTransfer' as const, [] as NftTransferDetails[]],
+    ['exchangeTransfer' as const, {} as UnsignedExchangeTransferRequest],
+    ['getAddress' as const, {} as any],
+    ['isRegisteredOffchain' as const, {} as any],
+  ])('when the user has been logged out - %s', (methodName, args) => {
+    beforeEach(() => {
+      passportEventEmitter.emit(PassportEvents.LOGGED_OUT);
+    });
+
+    it(`should return an error for ${methodName}`, async () => {
+      expect(async () => await passportImxProvider[methodName!](args))
+        .rejects
+        .toThrow(
+          new PassportError(
+            'User has been logged out',
+            PassportErrorType.NOT_LOGGED_IN_ERROR,
+          ),
+        );
+    });
```

Comment:
When asserting errors with custom types, like here with `PassportErrorType.NOT_LOGGED_IN_ERROR` jest doesn't actually assert this type is what we expect and only validates the message is the same. You could pass in any `PassportErrorType` here and the test will pass.

E.g. the following will pass the test

```
expect(async () => await passportImxProvider[methodName!](args))
        .rejects
        .toThrow(
          new PassportError(
            'User has been logged out',
            'Some type' as PassportErrorType,
          ),
        );
```

If you want to assert the type the only way I've found works is using a catch block to assert the type is what is expected. Note that we need to assert outside of the catch, otherwise we cant be sure that the catch was evaluated as the test will pass if the function doesn't throw.

```
let message = '';
let type = '';
try {
        await passportImxProvider[methodName!](args);
} catch (e: any) {
        message = e.message;
        type = e.type;
}

expect(message).toEqual('User has been logged out');
expect(type).toEqual(PassportErrorType.NOT_LOGGED_IN_ERROR);
```

Reply:
Thanks for the callout. I just moved this test over from `passportImxProvider.test.ts`, but probably a good opportunity to fix it up while we're in the area.